pkg/attribute/mutableBag.go

```diff
@@ -206,198 +203,209 @@ func (mb *MutableBag) Merge(bags ...*MutableBag) error {
-	for k, v := range attrs.DurationAttributes {
-		if log != nil {
-			log.WriteString(fmt.Sprintf("  updating duration attribute %s from '%v' to '%v'\n", dictionary[k], mb.values[dictionary[k]], v))
-		}
-		mb.values[dictionary[k]] = v
+	log(buf, "  setting duration attributes:\n")
+	for k, v := range attrs.Durations {
+		name, e = lookup(k, e, globalDict, messageDict)
+		log(buf, "    %s -> '%v'\n", name, v)
+		mb.values[name] = v
 	}
 
-	for k, v := range attrs.BytesAttributes {
-		if log != nil {
-			log.WriteString(fmt.Sprintf("  updating bytes attribute %s from '%v' to '%v'\n", dictionary[k], mb.values[dictionary[k]], v))
-		}
-		mb.values[dictionary[k]] = v
+	log(buf, "  setting bytes attributes:\n")
+	for k, v := range attrs.Bytes {
+		name, e = lookup(k, e, globalDict, messageDict)
+		log(buf, "    %s -> '%s'\n", name, v)
+		mb.values[name] = v
 	}
 
-	for k, v := range attrs.StringMapAttributes {
-		m, ok := mb.values[dictionary[k]].(map[string]string)
-		if !ok {
-			m = make(map[string]string, len(v.Map))
-			mb.values[dictionary[k]] = m
-		}
+	log(buf, "  setting string map attributes:\n")
+	for k, v := range attrs.StringMaps {
+		name, e = lookup(k, e, globalDict, messageDict)
+		log(buf, "  %s\n", name)
 
-		if log != nil {
-			log.WriteString(fmt.Sprintf("  updating stringmap attribute %s from\n", dictionary[k]))
+		sm := make(map[string]string, len(v.Entries))
+		for k2, v2 := range v.Entries {
+			var name2 string
+			var value2 string
+			name2, e = lookup(k2, e, globalDict, messageDict)
+			value2, e = lookup(v2, e, globalDict, messageDict)
+			log(buf, "    %s -> '%v'\n", name2, value2)
+			sm[name2] = value2
+		}
+		mb.values[name] = sm
+	}
 
-			if len(m) > 0 {
-				for k2, v2 := range m {
-					log.WriteString(fmt.Sprintf("    %s:%s\n", k2, v2))
-				}
-			} else {
-				log.WriteString("    <empty>\n")
-			}
+	if buf != nil {
+		glog.Info(buf.String())
+		pool.PutBuffer(buf)
+	}
 
-			log.WriteString("  to\n")
-		}
+	if e != nil {
+		return nil, e
+	}
 
-		for k2, v2 := range v.Map {
-			m[dictionary[k2]] = v2
-		}
+	return mb, nil
+}
 
-		if log != nil {
-			if len(m) > 0 {
-				for k2, v2 := range m {
-					log.WriteString(fmt.Sprintf("    %s:%s\n", k2, v2))
-				}
-			} else {
-				log.WriteString("    <empty>\n")
-			}
+func lookup(index int32, err error, globalDict []string, messageDict []string) (string, error) {
+	if index < 0 {
+		if -index-1 >= int32(len(messageDict)) {
+			return "", me.Append(err, fmt.Errorf("attribute index %d is not defined in the available dictionaries", index))
 		}
+		return messageDict[-index-1], err
 	}
 
-	if log != nil {
-		if log.Len() > 0 {
-			glog.Infof("Updating attribute bag %d:\n%s", mb.id, log.String())
-		}
+	if index >= int32(len(globalDict)) {
+		return "", me.Append(err, fmt.Errorf("attribute index %d is not defined in the available dictionaries", index))
 	}
 
-	return nil
+	return globalDict[index], err
+}
+
+func log(buf *bytes.Buffer, format string, args ...interface{}) {
```

Comment:
We'll need a `// nolint: interfacer` here, right? Also, occurs to me now that it'd be cleaner to declare this as an anonymous function inside of `GetBagFromProto` since it's only used there.